stdlib/random.gr

```diff
@@ -0,0 +1,176 @@
+  Memory.decRef(WasmI32.fromGrain(squares))
+  ret
+}
+
+// [NOTE] The PRNG algorithm seems to not work as well with small seeds, but
+// users are more likely to use small numbers for their seeds. Consequently,
+// we produce an actual seed xoring feeding the given seed with a
+// constant (large) number.
+
+/**
+ * Creates a new random number generator with the given seed.
+ *
+ * @param seed: The seed for the random number generator
+ * @returns The random number generator
+ */
+export let make = seed => {
+  { seed, counter: 0L, initialized: false }
+}
+
+/**
+ * Creates a new random number generator with a random seed.
+ *
+ * @param seed: The seed for the random number generator
+ * @returns The random number generator
+ */
+export let makeUnseeded = () => {
+  // TODO: Should we just .expect this result for UX's sake?
+  Result.map(seed => {
+    { seed, counter: 0L, initialized: false }
+  }, WasiRandom.randomInt64())
+}
+
+/**
+ * [Internal note]
+ * For low seed numbers, we sometimes need to churn through
+ * some iterations to start getting interesting numbers. Taking
+ * a cue from the API in https://pypi.org/project/squares-rng/ ,
+ * we churn through until we generate an int with a MSB of 1.
+ * Then, to avoid making all of the first generated numbers negative,
+ * we do another increment at the end.
+ */
+let checkInitialized = (random: Random) => {
+  if (!random.initialized) {
+    while (Int32.gt(Int32.clz(squares(random.counter, random.seed)), 0l)) {
+      incCounter(random)
+    }
+    // now that it's initialized, increment it again to make it a little more random
+    incCounter(random)
+    random.initialized = true
+  }
+}
+
+/**
+ * Generates a random 32-bit integer from the given random number generator.
+ *
+ * @param random: The random number generator to use
+ * @returns The randomly generated number
+ */
+export let nextInt32 = (random: Random) => {
+  checkInitialized(random)
+  let ret = squares(random.counter, random.seed)
+  incCounter(random)
+  ret
+}
+
+/**
+ * Generates a random 64-bit integer from the given random number generator.
+ *
+ * @param random: The random number generator to use
+ * @returns The randomly generated number
+ */
+export let nextInt64 = (random: Random) => {
+  checkInitialized(random)
+  let ret1 = Int64.fromNumber(
+    Int32.toNumber(squares(random.counter, random.seed))
+  )
+  incCounter(random)
+  let ret2 = Int64.fromNumber(
+    Int32.toNumber(squares(random.counter, random.seed))
+  )
+  incCounter(random)
+  Int64.lor(Int64.shl(ret1, 32L), ret2)
+}
+
+@disableGC
+export let rec gtU = (x: Int32, y: Int32) => {
+  let xv = WasmI32.load(WasmI32.fromGrain(x), 8n)
+  let yv = WasmI32.load(WasmI32.fromGrain(y), 8n)
+  let ret = WasmI32.gtU(xv, yv)
+  Memory.decRef(WasmI32.fromGrain(x))
+  Memory.decRef(WasmI32.fromGrain(y))
+  Memory.decRef(WasmI32.fromGrain(gtU))
+  ret
+}
+
+export let nextInt32InRange = (random: Random, low: Int32, high: Int32) => {
```

Comment:
Probably weird to hear me say this, but it could be worth it to implement this via low-level instructions.